----

### Proyecto 11

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

# Verificamos las primeras filas
for i, data in enumerate([data_0, data_1, data_2]):
    display(f"Región {i}:\n")
    display(data.info())
    display(data.head(), "\n")

'Región 0:\n'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

'\n'

'Región 1:\n'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

'\n'

'Región 2:\n'

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


None

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

'\n'

### 2

In [10]:

# pase de nuevo las librerias ya que no las estaba detectando 
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

def train_and_evaluate(data):
    # Seleccione las características (features) y el objetivo (target)
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    
    # Divide los datos en entrenamiento y validación (75% / 25%)
    X_train, X_val, y_train, y_val = train_test_split(
        features, target, test_size=0.25, random_state=42)
    
    # Cree y entrenamos el modelo de regresión lineal
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    # Realice las predicciones con los datos de validación
    y_pred = model.predict(X_val)
    
    # Calcule el volumen medio predicho y el RMSE
    mean_pred = np.mean(y_pred)
    rmse = mean_squared_error(y_val, y_pred, squared=False)
    
    # Imprimi los resultados para verificar que todo este en orden 
    print(f"Volumen medio predicho: {mean_pred:.2f}")
    print(f"RMSE: {rmse:.2f}")
    
    # Devuelve las predicciones y valores reales para usarlos luego
    return y_val, y_pred

# Ejecuta la función para las tres regiones
for i, data in enumerate([data_0, data_1, data_2]):
    print(f"Región {i}:")
    y_val, y_pred = train_and_evaluate(data)
    print()


Región 0:
Volumen medio predicho: 92.40
RMSE: 37.76

Región 1:
Volumen medio predicho: 68.71
RMSE: 0.89

Región 2:
Volumen medio predicho: 94.77
RMSE: 40.15



Con este paso nos podemos dar cuenta que a pesar que en la region 1 se producirian menos barriles que en otras regiones,es el mas preciso de los 3, por ende el que mas confianza da ya que realemnte muestra lo que puede ocurrir. o almenos hasta este momento.

### 3

In [13]:


# Umbral mínimo por pozo para evitar pérdidas (cada unidad vale $1000)
# 500,000 dólares / 1000 = 500 unidades mínimas por pozo
min_units = 500  

# Usamos la función del paso anterior para obtener las predicciones
# Esto llama a la función train_and_evaluate(data) y saca los valores predichos

_, y_pred_0 = train_and_evaluate(data_0)
_, y_pred_1 = train_and_evaluate(data_1)
_, y_pred_2 = train_and_evaluate(data_2)

# Calculamos el volumen medio predicho por región
mean_0 = np.mean(y_pred_0)
mean_1 = np.mean(y_pred_1)
mean_2 = np.mean(y_pred_2)

# promedios
print("Promedio de reservas predichas por región:")
print(f"Región 0: {mean_0:.2f} unidades")
print(f"Región 1: {mean_1:.2f} unidades")
print(f"Región 2: {mean_2:.2f} unidades\n")

# Comparamos con el umbral mínimo necesario
print(f"Para no tener pérdidas, cada pozo debe generar al menos {min_units} unidades.\n")

if mean_0 > min_units:
    print("Región 0 tiene reservas suficientes.")
else:
    print("Región 0 no cumple con el mínimo necesario.")

if mean_1 > min_units:
    print("Región 1 tiene reservas suficientes.")
else:
    print("Región 1 no cumple con el mínimo necesario.")

if mean_2 > min_units:
    print("Región 2 tiene reservas suficientes.")
else:
    print("Región 2 no cumple con el mínimo necesario.")


Volumen medio predicho: 92.40
RMSE: 37.76
Volumen medio predicho: 68.71
RMSE: 0.89
Volumen medio predicho: 94.77
RMSE: 40.15
Promedio de reservas predichas por región:
Región 0: 92.40 unidades
Región 1: 68.71 unidades
Región 2: 94.77 unidades

Para no tener pérdidas, cada pozo debe generar al menos 500 unidades.

Región 0 no cumple con el mínimo necesario.
Región 1 no cumple con el mínimo necesario.
Región 2 no cumple con el mínimo necesario.


### 4


In [ ]:


def train_and_evaluate(data):
    features = data[['f0', 'f1', 'f2']]
    target = data['product']
    
    X_train, X_val, y_train, y_val = train_test_split(
        features, target, test_size=0.25, random_state=42)
    
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    y_pred = model.predict(X_val)
    
    mean_pred = np.mean(y_pred)
    rmse = mean_squared_error(y_val, y_pred, squared=False)
    
    print(f"Volumen medio predicho: {mean_pred:.2f}")
    print(f"RMSE: {rmse:.2f}")
    
    return y_val, y_pred
# Parámetros económicos
unit_price = 4500       # Precio por unidad
investment = 100_000_000  # Inversión fija

# Guarda los resultados reales y predichos para cada región
results = []

for i, data in enumerate([data_0, data_1, data_2]):
    print(f"Procesando Región {i}...")
    
    # Obtenemos valores reales y predichos usando la función 
    y_val, y_pred = train_and_evaluate(data)
    
    # Construi DataFrame con valores reales y predichos
    df = pd.DataFrame({
        'real': y_val.reset_index(drop=True),
        'predicted': y_pred
    })
    
    # Ordene  por valor predicho descendente y seleccionamos los mejores 200 pozos
    top_200 = df.sort_values(by='predicted', ascending=False).head(200)
    
    # Calcule ganancia: suma del volumen real * precio unidad menos inversión
    profit = top_200['real'].sum() * unit_price - investment
    
    print(f"Ganancia Región {i}: ${profit:,.0f}\n")
    
    results.append(profit)

# Elegi la mejor región
best_region = np.argmax(results)
print(f"La mejor región para inversión es la Región {best_region} con una ganancia de ${results[best_region]:,.2f}")



Procesando Región 0...
Volumen medio predicho: 92.40
RMSE: 37.76
Ganancia Región 0: $33,591,411

Procesando Región 1...
Volumen medio predicho: 68.71
RMSE: 0.89
Ganancia Región 1: $24,150,867

Procesando Región 2...
Volumen medio predicho: 94.77
RMSE: 40.15
Ganancia Región 2: $25,985,718

La mejor región para inversión es la Región 0 con una ganancia de $33,591,411.14


### 5

In [8]:
def calcula_riesgo_y_ganancias_bootstrap(df_val, wells=200, sample_size=500, unit_price=4500, investment=100_000_000, n_bootstrap=1000):
    ganancias_bootstrap = []

    for _ in range(n_bootstrap):
        
        muestra = df_val.sample(n=sample_size, replace=True)
        
       
        top_200 = muestra.sort_values(by='predicted', ascending=False).head(wells)
        
      
        ganancia = top_200['real'].sum() * unit_price - investment
        ganancias_bootstrap.append(ganancia)

    
    ganancias_bootstrap = np.array(ganancias_bootstrap)
    ganancia_media = ganancias_bootstrap.mean()
    intervalo_95 = np.percentile(ganancias_bootstrap, [2.5, 97.5])
    riesgo_perdida = np.mean(ganancias_bootstrap < 0) * 100

    return ganancia_media, intervalo_95, riesgo_perdida


for i, df_val in enumerate(validation_data):
    media, intervalo, riesgo = calcula_riesgo_y_ganancias_bootstrap(df_val)
    print(f"Región {i}:")
    print(f" Ganancia promedio (bootstrap): {media:,.2f}")
    print(f" Intervalo de confianza 95%: {intervalo}")
    print(f" Riesgo de pérdida: {riesgo:.2f}%\n")



Región 0:
 Ganancia promedio (bootstrap): 4,156,827.38
 Intervalo de confianza 95%: [-733044.08938035 9084053.07476741]
 Riesgo de pérdida: 5.20%

Región 1:
 Ganancia promedio (bootstrap): 4,562,415.95
 Intervalo de confianza 95%: [ 831356.47111729 8636549.18378317]
 Riesgo de pérdida: 1.20%

Región 2:
 Ganancia promedio (bootstrap): 3,712,968.25
 Intervalo de confianza 95%: [-1649635.38613196  8681356.59812661]
 Riesgo de pérdida: 9.30%



El bootstrapping nos da a entender que La Región 1 es la que muestra la mayor ganancia media y un riesgo de pérdida del 1.20%, lo que la hace la opción más atractiva para invertir. a diferencia de las otras dos regiones donde tienen un riesgo de perdida un poco mas alto y las ganancias no son tan atractivas.
